### DISCUSSION TITLE
post multipart/form-data to stored procedure?

### DISCUSSION THREAD

**Discussion by SupaFuzz** [PostgREST/postgrest#2259] [Q&A] [answered]
Hi Everyone,

I know this is a long shot, but I'm wondering if anyone has ever hacked on postgrest to get this kind of thing working?

Long story short is that I want to build a REST service on postgrest to receive webhook calls from an external system (BMC Remedy not that it really maters). I'd like to create a table with a `jsonb` column and just insert the verbatim JSON payload into it. This actually seems to be something that would work as the webhook calls most often have the`Content-Type: application/json` so in thise cases, I suspect I can get that much to work.

The complication is that sometimes the external system will attempt to send files as well as a JSON payload in the body of the webhook call, and in those cases, it will use `Content-type: multipart/form-data` with `BASE64` encoded files in the body.

If at all possible I'd like to pull those out of the body as well and insert them into `bytea` columns. 

The thought occurs that creating the stored procedure in a procedural language like `PL/Perl` or maybe even `plv8` would likely provide the capability of parsing out complicated request bodies like this, but is there any way to get postgrest to just pass the entire request body into the stored procedure to do something like that?

Anyone bark up this tree before and want to share some hints?

thanks everyone, 

-Amy

**Comment by steve-chavez** [comment, marked as answer]
Hey Amy,

> This actually seems to be something that would work as the webhook calls most often have the Content-Type: application/json so in thise cases, I suspect I can get that much to work.

Yes, this should work. We have the convention that a function with a single unnamed json param will get the raw payload:

https://postgrest.org/en/stable/api.html#calling-functions-with-a-single-unnamed-parameter

> The complication is that sometimes the external system will attempt to send files as well as a JSON payload in the body of the webhook call, and in those cases, it will use Content-type: multipart/form-data
> If at all possible I'd like to pull those out of the body as well and insert them into bytea columns.

The above also works for single unnamed bytea params. However, that only works with `Content-Type: application/octet-stream`.

Do you have a proxy in place that can change the `Content-type: multipart/form-data` header to `Content-Type: application/octet-stream`?

**Reply by SupaFuzz** [reply]
Interesting approach. I could probably setup a proxy. IIRC the docs suggest using ngix for hardening anyhow, so I could play with that.

Just to be sure I'm understanding you correctly, you're saying if I setup a stored proc that takes a single `bytea` param and the request to it specifies `Content-Type: application/octet-stream`, that the stored proc will receive the complete raw request body as the single argument?

**Reply by steve-chavez** [reply]
Yes, correct.

**Reply by prgueza** [reply]
Hey all! 

I've been trying to figure out a solution quite similar to the one @steve-chavez proposed here and I'm having some trouble to discern between the part of the form that corresponds to fields and the file per se. The multipart form looks something like this:

```
----------------------------194211810012555286961105
Content-Disposition: form-data; name="user_data"

This is some user value
----------------------------194211810012555286961105
Content-Disposition: form-data; name="file_data"; filename="test.txt"
Content-Type: text/plain

some content

----------------------------194211810012555286961105--
```

And I would like to store the `user_data` into a column and do something else with the `file_data` (in my case send it to S3, but in yours it would be to store it into a different column (bytea)).

@SupaFuzz, have you managed to extract this information from the raw bytea that the function receives?

Thank you in advance!

**Reply by SupaFuzz** [reply]
@prgueza Unfortunately I had to abandon work on the project before I got it nailed down. I'd gotten as far as using NGINX to munge the content-type to get it all to dump the entire request into a stored procedure. I'd been experimenting with plv8 to try and handle manually parsing the raw request in the procedure. 

I think I could've made that work but unfortunately just have had to switch focus and never have gotten back around to working on it. The roadblock I was working on when I left off is that the content separator string was specified in the header. Meaning to get the delimiter to parse the raw request body, I needed to parse the header too. I think that was possible, I was playing around with how to extract that key from the header when I had to walk away.

best of luck! 
if you figure it out please post the solution!

**Reply by prgueza** [reply]
Thanks for replying @SupaFuzz!

I couldn't spend much time trying to figure out how to make it work with multipart forms, so I ended up using a `rpc` with a `json` body and sending the file data as a `base64` encoded string as I've seen proposed in some other issues (with the drawback of increasing the size by 33%).

What you say about reading the separator from the headers sounds pretty good, PostgREST talks about this [here](https://postgrest.org/en/stable/api.html#http-context), I guess you've read it already (and maybe even tried it out), hopefully the next person to come can pick it up from here!

I'll leave here a simplified version of the function I used in case it helps someone out, but it would be cool to be able to handle multipart form data from PostgREST at some point 😄 

```
CREATE OR REPLACE FUNCTION api.post_dataset("name" TEXT, "description" TEXT, "data" TEXT)
  RETURNS JSON
  SECURITY DEFINER
  LANGUAGE plpgsql
AS $FUNCTION$

  DECLARE
  
    -- ARGS
    _name TEXT := name;
    _description TEXT := description;
    _data TEXT := data;
    -- VARIABLES
    _object_key TEXT;

  BEGIN  

    SELECT 
    CONCAT_WS('/', 'dataset', TO_CHAR(NOW(), 'YYYY_MM_DD'), public.slugify(_name, 'dataset'))
    INTO _object_key;

    PERFORM storage.s3_write_file(_object_key, DECODE(_data, 'base64')); -- Decoded back to binary before sending it to S3

    RETURN JSON_BUILD_OBJECT('object_key', _object_key);

  END;

$FUNCTION$;
```
